### PR TITLE
Use node IP in advertised.listeners

### LINKED ISSUE #900: NodePort enhancement: use Node IP in advertized.listeners when hostnameOverride and nodePortExternalIP are not set
**Is your feature request related to a problem? Please describe.**
Reaching the Kafka cluster with IP address would be useful. Currently with hostnameOverride a list of ip addresses and hostnames has to be maintained in /etc/hosts.

**Describe the solution you'd like to see**
when hostnameOverride and nodePortExternalIP are not set use the node ip in advertized.listeners configuration.

**Describe alternatives you've considered**

**Additional context**
https://banzaicloud.com/blog/kafka-external-access/#future-nodeport-enhancements

Future NodePort enhancements[ 🔗︎](https://banzaicloud.com/blog/kafka-external-access/#future-nodeport-enhancements)
Currently, either hostnameOverride or nodePortExternalIP, or both, have to be specified for external listeners in the NodePort access method. In future versions, if these are not set, the node IP of the node where the broker pod is scheduled will be used in the advertized.listeners broker configuration.

### PR DESCRIPTION
**Pull request by balassai**
| Q               | A      |
| --------------- | ------ |
| Bug fix?        | no|
| New feature?    | yes |
| API breaks?     | no|
| Deprecations?   | no|
| Related tickets | fixes #900  |
| License         | Apache 2.0 |

### What's in this PR?
Use node IP in advertised.listeners when hostnameOverride and nodePortExternalIP are not set

### Additional context
When "hostNameOverride" and brokerConfig.nodePortExternalIP are empty and NodePort access method is selected the NodePortNodeAdddressType defines the external listener Address that shall be used in the advertised.listeners property. 
If NodePortNodeAddressType is not provided the first non empty address is used in the following order:
1. ExternalIP
2. InternalIP

### Checklist

- [ x] Implementation tested
- [x ] Error handling code meets the [guideline](https://github.com/banzaicloud/developer-guide/blob/master/docs/coding-style/error-handling-guide.md)
- [x ] Logging code meets the guideline
- [x] User guide and development docs updated (if needed)

### CODE REVIEW

**Review by bartam1 — CHANGES_REQUESTED** [review]

**Comment by bartam1** [comment]
Thank you for your effort to do this!

File: pkg/resources/kafka/kafka.go [hunk 1]
```diff
@@ -1218,6 +1225,48 @@ func (r *Reconciler) getK8sAssignedNodeport(log logr.Logger, eListenerName strin
 	return nodePort, nil
 }
 
+func (r *Reconciler) getK8sAssignedNode(log logr.Logger, brokerId int32, nodeAddressType string) (string, error) {
```
Comment (bartam1) [inline pkg/resources/kafka/kafka.go:1228]: We can use more descriptive function name like: getK8sAssignedNodeAddress

File: api/v1beta1/kafkacluster_types.go [hunk 2]
```diff
@@ -191,6 +191,13 @@ type BrokerConfig struct {
 	// type service to expose the broker outside the Kubernetes cluster. Also, when "hostnameOverride" field of the external listener is set
 	// it will override the broker's external listener advertise address according to the description of the "hostnameOverride" field.
 	NodePortExternalIP map[string]string `json:"nodePortExternalIP,omitempty"`
+	// When "hostNameOverride" and brokerConfig.nodePortExternalIP are empty and NodePort access method is selected
+	// the NodePortNodeAdddressType defines the external listener Address that shall be used in the advertised.listeners property.
+	// https://kubernetes.io/docs/concepts/architecture/nodes/#addresses
+	// If NodePortNodeAddressType is not provided the first non empty address is used in the following order:
+	// 1. ExternalIP
+	// 2. InternalIP
```
[the remark this answers is not in the document]
Comment (pregnor) [reply api/v1beta1/kafkacluster_types.go:199]: In that case please also update the describing comment accordingly.

**Review by pregnor — COMMENTED** [review]
Aside from Márton's comments I didn't find anything else.
Thanks for the contribution.

**Comment by panyuenlau** [comment]
Hey @balassai, since this PR has been opened for a while, can you provide some updates on this?

**Comment by Jcodelove** [comment]
A very useful function. I hope to see it on the main branch immediately！！

**Comment by bartam1** [comment]
Im going to fix this PR and we can merge it.

**Comment by bartam1** [comment]
Merged into the master
cc @Jcodelove

**Pull request closed without merging**